services/src/main/java/io/scalecube/services/RemoteServiceInstance.java

```diff
@@ -78,39 +80,39 @@ private Type extractGenericReturnType(Method method) {
     }
   }
 
+  // Listen response
+  private void handleReply(Message message) {
+    String correlationId = message.correlationId();
+    ResponseFuture future = futures.remove(message.correlationId());
+    if (future != null) {
+      if (message.header("exception") == null) {
+        future.complete(message);
+      } else {
+        LOGGER.error("cid [{}] remote service invoke respond with error message {}", correlationId, message);
+        future.completeExceptionally(message.data());
+      }
+    }
+  }
+
   private CompletableFuture<Object> futureInvoke(final Message request, Function<Message, Object> fn) throws Exception {
-    final CompletableFuture<Object> messageFuture = new CompletableFuture<>();
-
-    final String correlationId = "rpc-" + generateId();
-
-    Message requestMessage = composeRequest(request, correlationId);
-    // Listen response
-    this.cluster.listen()
-        .filter(message -> correlationId.equals(message.correlationId()))
-        .first().subscribe(message -> {
-          if (message.header("exception") == null) {
-            messageFuture.complete(fn.apply(message));
-          } else {
-            LOGGER.error("cid [{}] remote service invoke respond with error message {}", correlationId, message);
-            messageFuture.completeExceptionally(message.data());
-          }
-        });
 
-    // check that send operation completed successfully else report an error
+    ResponseFuture responseFuture = new ResponseFuture(fn);
+    futures.putIfAbsent(responseFuture.correlationId(), responseFuture);
```

Comment:
What about handling methods which return `void`, you will put objects into this map but those methods do not send response so it would be never removed and cause a memory leak. 

Also I don't see where this map of futures is cleared for the case if service response message was lost. It seems that in such case the future will be kept in this map forever so also cause memory leak.